pkg/config/remote/service/service.go

```diff
@@ -907,6 +911,40 @@ func filterNeededTargetFiles(neededConfigs []string, cachedTargetFiles []*pbgo.T
 	return filteredList, nil
 }
 
+func (s *CoreAgentService) apiKeyUpdateCallback() func(string, any, any) {
+	return func(setting string, oldvalue, newvalue any) {
+		if setting != "api_key" {
+			return
+		}
+
+		newKey, ok := newvalue.(string)
+		s.Lock()
+		defer s.Unlock()
+
+		if ok {
+			s.api.UpdateAPIKey(newKey)
+		}
+
+		// Verify that the Org UUID hasn't changed
+		orgUUID, err := s.uptane.StoredOrgUUID()
+		if err == nil {
+			// If we don't have a stored UUID, skip this check
+			newOrgUUID, err := s.api.FetchOrgData(context.Background())
+			if err == nil {
+				// If we weren't able to get the new org UUID, again skip
+				if orgUUID != newOrgUUID.Uuid {
+					log.Errorf("Error switching API key: new API key is from a different organization")
+					oldKey, ok := oldvalue.(string)
+					if ok {
+						s.api.UpdateAPIKey(oldKey)
```

Comment:
The request for more information was just to make it more obvious to users (and first level support investigating FLARE's) -why- we're failing other than just logging a 403 in order to prevent escalations to RC. Recall that we checked the requirements and we are allowed to fail here, we just want it more clear. Continuing to use the old key doesn't feel correct.